Add support for bulk transfers

xobs opened the pull request:
Add support for bulk transfers, which are present in V2.

This greatly increases the transfer rate, with performance increases of 3-4x seen in light testing.

~~Unfortunately legacy HID transfers are slower as a result of this patch, for reasons which are currently unclear.~~ Edit: Fixed in a followup patch.

devanlai committed:
Add USB2.1 binary object store support

devanlai committed:
Add WinUSB / Microsoft OS Descriptors 2.0 support

Targets with WinUSB compatible bootloaders can now detach and upgrade
firmware over DFU without installing drivers on Windows 10

xobs committed:
usb: use IRQ for usb

Handle USB traffic in an interrupt. This doesn't change download speed
much, but is necessary to get bulk endpoints working.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
support bulk communications

Support communicating with the host via bulk communications. This
works alongside HID, and both should be usable at the same time.

Operation performance has increased 3-4x with this patch.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
hid: get hid working again

It's slow, but it appears to work.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
DAP42: re-enable watchdog

Enable the watchdog now that we don't need to debug it anymore.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
dap42: enable bulk mode on DAP42

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
SW_DP: fix off-by-one in reference implementation

The original SWD sequence has an off-by-one error that causes
all returned values to be shifted. This isn't an issue for most
software, since they tend to use `SWD_Sequence()` as a write-only
primitive and don't care about the return value.

The Black Magic Debug Project uses this for additional checks,
which works around this bug if it's detected. For details, see:
https://github.com/blackmagic-debug/blackmagic/blob/31d30ada18862cdf44ae7ed798325cb7366d05e0/src/platforms/hosted/dap_command.c#L428-L458

This bug is present in reference code from ARM. If this file is
updated to a newer build, please ensure this fix is carried
forward, or ensure the newer version no longer has this bug.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs committed:
app: fix hid speed and performance

HID was previously broken. It was slow, and returned invalid data.
Fix HID so that it works like it did before Bulk was added. This
gives it the same performance as before.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs commented:
I managed to fix HID so that it works like it did before the rework.

For performance comparisons, here's the results of using `compare-sections` to check that a program matches what is on the target. This is better than measuring flash performance, since reads are usually only limited by the debugger speed.

Via Black Magic Debug Application:

Bulk:
```
bmd_crc32: 0x08000000+192 -> 4ms
bmd_crc32: 0x080000c0+52320 -> 701ms, 72 KiB/s
bmd_crc32: 0x0800cd20+15664 -> 204ms, 74 KiB/s
bmd_crc32: 0x08010a50+1028 -> 16ms, 62 KiB/s
```

HID:
```
bmd_crc32: 0x08000000+192 -> 14ms
bmd_crc32: 0x080000c0+52320 -> 2008ms, 25 KiB/s
bmd_crc32: 0x0800cd20+15664 -> 598ms, 25 KiB/s
bmd_crc32: 0x08010a50+1028 -> 44ms, 22 KiB/s
```

HID (before the patch):
```
bmd_crc32: 0x08000000+192 -> 13ms
bmd_crc32: 0x080000c0+52320 -> 2007ms, 25 KiB/s
bmd_crc32: 0x0800cd20+15664 -> 604ms, 25 KiB/s
bmd_crc32: 0x08010a50+1028 -> 43ms, 23 KiB/s
```

devanlai commented:
Wow, those results are quite impressive. Thanks for taking the time to implement this.

I'll have to take some time to remember how the code worked (or aspired to work at least) and make sure everything still makes sense with your changes to the buffer management.

Aside from that, my main concern would be making sure that the compile-time resource constraint assertions are still up to date. In particular, it looks like the new bulk endpoints are not represented in the packet memory area calculations in `composite_usb_conf.c`.

xobs committed:
composite_usb_conf: add bulk to packet memory calculations

Packet memory is a finite resource. This is memory where packet data
is buffered as it goes to the host, or as packets are received from
the host.

Previously, bulk endpoints were not accounted for. This didn't cause
any observable issues because there is enough memory. Add the check
for bulk data in case this is ported to a device that has a packet
memory area that's too small.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs commented:
I missed that. Good catch. I've added it to the check in https://github.com/devanlai/dap42/pull/23/commits/19e5021b9d9d41a0b0258a97b563d35f74e1237d

I also made HID optional in the calculation.

devanlai commented:
Alright, I think I've reloaded some of the context around the USB handling now.

In no particular order, my general thoughts:

- I like how you've slimmed it down to one circular buffer for the incoming DAP commands. It was definitely unnecessary to have dedicated slots for both the request and the response, given that they only briefly both exist simultaneously.
- I can believe that under normal testing, sending the response over the bulk interface immediately will not block because the host is likely to poll/drain the bulk interface faster than we can execute DAP commands and output the responses.
- I would feel better if it always put the response into the circular buffer to handle the edge case where requests come in faster than the responses are drained.
- In the unlikely event that requests come in approximately as fast as they're processed, it might be possible to sit in the command processing loop in `DAP_app_update()` indefinitely, starving the rest of the superloop and triggering the watchdog. I'd be curious if there's a performance hit to capping the number of commands processed to per invocation to at most one or two to guard against that.
- I forget whether zero-length packets are actually required for CMSIS-DAP. They're definitely needed for CDC-ACM which can carry arbitrary unframed data, but it may be unnecessary for CMSIS-DAP where it can be understood at the protocol level that each response gets its own standalone packet.
- Not related to any of your changes, but I recall now that there is a race condition between trying to fill the HID IN endpoint buffer from `DAP_app_update()` vs the USB callback for the HID IN endpoint trying to refill it from the buffer after the endpoint is polled. I see that your scheme for the bulk endpoint of only populating the endpoint from from `DAP_app_update()` avoids that.

I'll have to think about it some more and dust off my dev boards to do some experiments.

xobs commented:
Regarding the ZLP requirement -- it definitely was required when I did it. If you send 64 bytes on a USB FS link, then the host must ask for the rest of the buffer. You indicate that there is no more data by sending a ZLP. I didn't get any data before I added that.

I believe it's the case for Interrupt endpoints that they can only transmit at most 64 bytes, so there's no need for a ZLP there.

Perhaps you're right in that the clients I've worked with always do an OUT followed immediately by an IN, and I can definitely see the case where they pipeline multiple OUT requests, especially for very long sequences. I'll look into pipelining bulk requests. I left that bomb in there if the IN request fails, and that does make me a bit uncomfortable.

There's also the separate issue in that with the bulk endpoints, we probably should add the Microsoft descriptors. I'll work on that after I get pipelined bulk transfers working.

devanlai commented:
> Regarding the ZLP requirement -- it definitely was required when I did it. If you send 64 bytes on a USB FS link, then the host must ask for the rest of the buffer. You indicate that there is no more data by sending a ZLP. I didn't get any data before I added that.
> 
> I believe it's the case for Interrupt endpoints that they can only transmit at most 64 bytes, so there's no need for a ZLP there.

I think you're right here - now that I think about it more, it's up to the host to specify how many bytes they're expecting up front, and if they don't always know in advance, they have to rely on ZLPs to delimit the end. It might be possible to infer the response size from the request, but it's probably safest for them to rely on ZLPs for maximum compatibility.

> There's also the separate issue in that with the bulk endpoints, we probably should add the Microsoft descriptors. I'll work on that after I get pipelined bulk transfers working.

I have an [old branch](https://github.com/devanlai/dap42/tree/winusb-wip) that was adding the Microsoft descriptors for the DFU interface together with some webusb stuff. It should be possible to extract just the USB 2.1 BOS and MS OS 2.0 descriptor functionality and adapt the descriptors for just CMSIS-DAP 2.0 instead. (And as long as the webusb landing URL functionality is excluded, we won't go over the hardcoded limit of at most 4 control callbacks).

xobs committed:
bulk: support pipelining bulk IN transfers

Support having more than one bulk response in flight at a time. A host can
make multiple requests without reading the response. In this case, store it
in the ring buffer just like we do with HID.

As part of this, it was discovered that the ZLP handling was incorrect, and
the feature was removed entirely.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs commented:
I'm not right. ZLP was completely broken. I've removed it, and added support for pipelining.

I put together this completely awful gist that helped me to trigger various cases. It works now: https://gist.github.com/xobs/1bc85cbe7da8aef31c3b5443d752f2cf

I can have a look at that patch for winusb.

xobs commented:
Actually, I think winusb should go in a separate PR. I'm happy with this for now, given it passes the tests I've thrown at it.

I'll work on winusb separately. Was there any reason you didn't merge that branch? Do you recall what issues you ran into? I realize that was nine years ago!

devanlai commented:
> Actually, I think winusb should go in a separate PR. I'm happy with this for now, given it passes the tests I've thrown at it.
> 
> I'll work on winusb separately. Was there any reason you didn't merge that branch? Do you recall what issues you ran into? I realize that was nine years ago!

Yeah, it makes sense to make the winusb changes in a separate PR. I'll try to do some testing on the bulk endpoints this weekend.

There wasn't anything technically wrong with that branch, it just didn't seem like it was adding much to make the DFU runtime for switching into the bootloader slightly easier to access via winusb/webusb, especially since the way I wanted to structure it (separate class callbacks for webusb and winusb) was running over the control callback limit.

xobs commented:
I see. Then with the addition of bulk endpoints, it becomes more important to allow Windows to take advantage of the new interface without having to resort to Zadig, whereas that wasn't an issue before.

devanlai committed:
Normalize DAP processing

Use a single set of head/tail pointers regardless of the command source
Throw a compile-time error if neither HID nor bulk endpoints are enabled for DAP
Process at most one command each time through the loop
Avoid racing with the USB interrupt when sending reports;
only manually move reports from the queue to USB when the
USB IN endpoint is idle
Only transfer the actual response length when sending HID
responses

devanlai committed:
Disable USB HID GET_REPORT request

devanlai commented:
It looks like I'll have to do some more testing and debugging on my side:
1. Unrelated to your changes (even on the mainline branch), I'm seeing issues with the HID CMSIS-DAP interface timing out on Windows, so I'll have to get my proper development environment setup again.
2. My STLink/v2 clone dongles aren't enumerating correctly with the bulk interface enabled, so I'll have to dig into that.

I've also started a branch at [cmsis-dap-bulk](https://github.com/devanlai/dap42/tree/cmsis-dap-bulk) that aligns things a bit more closely with how I had originally planned to manage the USB interrupt handling. It remains to be seen whether this is a bad strategy with hidden race conditions.

xobs commented:
1. Strange. I haven't checked Windows at all so I haven't seen it. Do you have suggestions as to how to trigger those conditions?
2. I have o STM32F103 harware. Do you have recommendations on boards to get?

I'm working to gain access to a USB Beagle. I really should have bought my own by now, what with how useful they are.

I like the changes to buffers, assuming they work. The one thing I'd change is to mark `process_head` as `volatile` since it's updated both from an interrupt and a non-interrupt context. Otherwise the optimizer has a tendency to get clever.

devanlai commented:
Good point about `process_head`. I was thinking it would be okay since even though it's read from both an interrupt and non-interrupt context, it's only updated from a non-interrupt context, but that's probably still enough leeway for the optimizer to do something iffy.

On Windows I'm seeing it time out immediately on the very first CMSIS-DAP command to query it, so I imagine that will hopefully be straightforward to debug once I get all the tools in place.

For STM32F103 hardware, I don't have any specific recommendations - for the most popular form fact, there are many knockoff "mini" STLink/v2 dongles, but even before the era of the supply chain shortage, vendors have been dropping in STM32-like replacement chips, so they're a bit hard to source. And even then, the SWD pads are inconvenient to get to and the layout isn't standard, hence why I haven't gone in and hooked up the debugger yet.

I'm sure it'll be something obvious once I take a look at a dev board.

xobs commented:
Well, I've got some STM32F103 boards on order from Aliexpress anyway. They might take a week to get here, though.

For context: I like designing boards, and with an [STM32F042 costing only slightly more](https://jlcpcb.com/partdetail/Stmicroelectronics-STM32F042G6U6/C961597) than a Cortex-M header, and using less board space, I've added it to my bag of tricks and have started to use it on boards I make now. No crystal, gives UART and SWD (or JTAG!), takes up just over 4x4 mm with passives, and works with USB-C. So `dap42` is the perfect solution.

xobs commented:
I got ahold of a USB Beagle and got myself an STM32F103. I'll have a look at it tomorrow if you haven't gotten it working by then.

devanlai committed:
Fix f103 usb interrupts

The USB OTG interrupt is not applicable for our hardware

devanlai commented:
Good news, I worked out that I had picked out the wrong interrupt line for the wrong USB peripheral on the STM32F103 and it seems to enumerate and work fine under Linux after that fix. I guess I should've been a bit more mistrustful of untested code from seven years ago 😅.

I'll still have to do some more testing, but it should be easier now that at at least some of my debuggers are in a functional state.

xobs commented:
I was actually just tracking that down right now. I had just noticed that the interrupt wasn't firing.

Anyway, since I have the Beagle hooked up and it's running under Windows, I'll take a look at adding the winusb descriptors in.

xobs commented:
I also have a small patch to add JTAG support that's working without any issues. That can also be in a separate patch.

xobs committed:
winusb: advertise bulk endpoint when available

When the bulk endpoint is available, advertise it as being available by
reporting it in the BOS. This enables Windows users to use the new bulk
endpoint without resorting to using Zadig to manually install WINUSB.

Signed-off-by: Sean Cross <sean@xobs.io>

xobs commented:
I updated the patch to add winusb support. It now attaches to winusb.sys for both the DFU endpoint (if available) as well as the DAP bulk endpoint (if configured).

xobs commented:
By the way, newer versions of libopencm3 have functions such as `microsoft_os_register_descriptor_sets()` and `usbd_register_bos_descriptor()` so we wouldn't have to duplicate functionality in DAP42. I can use those, but they would require copying GPLv3 code from libopencm3 into DAP42 and I'm assuming you want to avoid doing that.

devanlai commented:
> By the way, newer versions of libopencm3 have functions such as microsoft_os_register_descriptor_sets() and usbd_register_bos_descriptor() so we wouldn't have to duplicate functionality in DAP42. I can use those, but they would require copying GPLv3 code from libopencm3 into DAP42 and I'm assuming you want to avoid doing that.

I'm not strongly opposed to incorporating GPLv3 code, but it also looks like those changes haven't landed in the master branch of libopencm3 yet. I'd be happy to revisit this along with updating to a newer version of libopencm3 in the future.

> I also have a small patch to add JTAG support that's working without any issues. That can also be in a separate patch.

Sure, the pinout is somewhat limited on the official hardware, but I'd be happy to take a look in another PR.

After doing some additional testing on Windows and looking through the openocd source code, I think it was a mistake on my part to change the HID interface to stop padding out the responses. I'm not sure if it's a platform difference or a version difference (the joy of openocd builds that don't report the commit sha), but at least with my Windows openocd binary, it looks like it tries to request a full 64 byte transfer and it times out when it only gets the minimal 3 byte response required for the `DAP_INFO` command.

After backing out that change, I'm no longer seeing any issues on Windows or Linux with openocd. I still have some further regression testing to do and some work to create new build variants for the STM32F103 platform which has to choose between the HID vs bulk interface, but I'll be merging this momentarily after applying that patch and fixing up the volatile flag on `process_head`.

Thanks for putting in the work to push this forward!
